### PR TITLE
 [ModLog] Use custom scopes for ModLog Config

### PR DESCRIPTION
**Pull request by Tobotimus**
Modlog is currently the biggest culprit for seriously large documents in the MongoDB backend, since it stores all cases as nested dicts in the guild scope. So, for example, on the Fortnite server, the guild document for @Kowlin's bot has exceeded 8MB. 

This PR gives each case its own document. It also does the same for casetypes. Not only does it remove the possibility of the document exceeding the maximum size in MongoDB, it's also just more efficient for all backends.

#### Config Bug
These changes highlighted a bug in Config, which I've fixed here but will probably open a new PR for: when a custom scope uses more than one primary key, but does this with a partial primary key: `conf.custom(SCOPE, primary_key_1).all()`, the defaults will be mixed in with the top-level of the returned dict, even though the actual documents are nested further in.

#### General ModLog improvements
Fixed a bunch of type-hints, and also added more support for when an object related to a case (user, moderator, channel etc.) can't be found (because it was deleted or something rather)

### CODE REVIEW

**Comment by Kowlin** [comment]
PR seems to work, however I think that there is a migration failur, it cannot reference to old mod cases within the ``casesfor`` command, or continue its old count.

**Comment by Tobotimus** [comment]
Thanks, will take a look at it later this afternoon.

**Comment by Tobotimus** [comment]
This PR now depends on #2795

File: redbot/__main__.py [hunk 1]
```diff
@@ -116,8 +116,9 @@ def main():
     red.add_cog(CogManagerUI())
     if cli_flags.dev:
         red.add_cog(Dev())
+    loop = asyncio.get_event_loop()
```
Comment (mikeshardmind) [inline redbot/__main__.py]: Already assigned in line 110, no need to refetch the loop here.

Comment (Tobotimus) [reply redbot/__main__.py]: Oops! Thanks, that one on line 110 was added in a merge from V3/develop

**Review by Kowlin — APPROVED** [review]
PR is working as intended. Data is migrated over to custom handler and deleted from the ``modlog.guild`` as it should. Pulling back on old cases works, and continuity in modlog is working as intended.

**Pull request merged**